stew/io2.nim

```diff
@@ -0,0 +1,1007 @@
+  ##
+  ## Returns number of bytes read from file descriptor.
+  when defined(posix):
+    if len(data) > 0:
+      while true:
+        let res = read(cint(handle), unsafeAddr data[0], csize_t(len(data)))
+        if res == -1:
+          let errCode = ioLastError()
+          if errCode == EINTR:
+            continue
+          else:
+            return err(errCode)
+        else:
+          return ok(uint(res))
+    else:
+      return ok(0)
+  elif defined(windows):
+    if len(data) > 0:
+      var lpNumberOfBytesRead = 0'u32
+      let res = readFile(uint32(handle), unsafeAddr data[0], uint32(len(data)),
+                         lpNumberOfBytesRead, nil)
+      if res != 0:
+        ok(lpNumberOfBytesRead)
+      else:
+        err(ioLastError())
+    else:
+      ok(0)
+
+proc readFile*(handle: IoHandle,
+               data: var openarray[char]): IoResult[uint] {.inline.} =
+  ## Reads ``len(data)`` characters from file descriptor ``handle`` and store
+  ## this characters to ``data``.
+  ##
+  ## Returns number of bytes characters read from file descriptor.
+  readFile(handle, data.toOpenArrayByte(0, len(data) - 1))
+
+proc writeFile*(pathName: string, data: openarray[byte],
+                createMode: int = 0o644): IoResult[void] =
+  ## Opens a file named ``pathName`` for writing. Then writes the
+  ## content ``data`` completely to the file and closes the file afterwards.
+  ##
+  ## If file is not exists it will be created with permissions mask
+  ## ``createMode`` (default value is 0o644).
+  ##
+  ## If file is already exists, but file permissions are not equal to
+  ## ``createMode`` procedure will change permissions first and only after
+  ## success it will write data to file.
+  if fileAccessible(pathName, {AccessFlags.Find, AccessFlags.Write}):
+    let permissions = ? getPermissions(pathName)
+    if permissions != createMode:
+      ? setPermissions(pathName, createMode)
+  let flags = {OpenFlags.Write, OpenFlags.Truncate, OpenFlags.Create}
+  let handle = ? openFile(pathName, flags, createMode)
+  var offset = 0
+  while offset < len(data):
+    let res = writeFile(handle, data.toOpenArray(offset, len(data) - 1))
+    if res.isErr():
+      # Do not care about `closeFile(handle)` error because we already in
+      # error handler.
+      discard closeFile(handle)
+      return err(res.error)
+    else:
+      offset = offset + int(res.get())
+  ? closeFile(handle)
+  ok()
+
+proc writeFile*(pathName: string, data: openarray[char],
+                createMode: int = 0o644): IoResult[void] {.inline.} =
+  ## Opens a file named ``pathName`` for writing. Then writes the
+  ## content ``data`` completely to the file and closes the file afterwards.
+  ##
+  ## If file is not exists it will be created with permissions mask
+  ## ``createMode`` (default value is 0o644).
+  writeFile(pathName, data.toOpenArrayByte(0, len(data) - 1), createMode)
+
+proc readAllFile*(pathName: string, blockSize = 16384'u): IoResult[seq[byte]] =
```

Comment:
perhaps `readAllFile(name: string, buf: var seq[byte): IOResult[void]` to stay in line with the other `readFile`? this overload could then stay around as well.

also, I'm not sure I understand the blocksize thing - why not read the file size first, then create a seq of that size, then read the data into it in a single read? this impl reallocates and copies the contents leading to basically `O(N^2)` overhead